PR for #50: controls economic shocks

gabrieleborg opened the pull request:
As per our discussion, I am sending this task to PR. I updated the code so that the baseline model presented in column 1 of the tables with economic controls is the one with _linear trend_ only.

santiagohermo reviewed (commented):
Thanks @gabrieleborg! It's really satisfying to see that the effect holds up to so many controls haha. I made several inline comments, mostly focused on making some parts of the code more readable. General comments:

- The script `analysis/first_differences/code/first_differences.do` is waaay too long. I recommend splitting it in two: `estimate_base.do` and `estimate_heterogeneity.do`, or something like that

- Some tables produced by `first_differences.do` had wrong formatting of labels. I fixed this. I also set the number of observations to show up with a thousands delimiter (","). To do this I changed the observations format from `%9.0g` to `%9.0gc`. We should implement this in other scripts as well (fyi @diegogentilepassaro)

- README of `raw/bps` is missing.

- `make.py` was missing in `base/bps/code` (code looks good). I added it and committed a re-run of this step

- In `base/qcew`, `make.py` had not been update with the new program `make_county_month_panel.do`. I added it and re-ran the folder.

I did not commit a new run of `/first_differences`. I assume you will do after implementing some comments!

gabrieleborg commented on raw/lau/code/build.py:
```diff
@@ -0,0 +1,21 @@
+#!/usr/bin/env python3
```
I ended up not using this data. I deleted the folder

gabrieleborg commented on base/qcew/code/make_county_month_panel.do:
```diff
@@ -0,0 +1,100 @@
+program select_sectors
+    syntax, ownership(str) industry(str)
+
+    local n : word count `ownership'
+    local first_owner: word 1 of `ownership'
+    local owner_list `"ownership=="`first_owner'""'
+    
+    if `n'>1 {
+        forval x = 2/`n' {
+            local this_owner: word `x' of `ownership'
+            local this_owner_list `" | ownership== "`this_owner'""'
+            local owner_list `"`owner_list' `this_owner_list'"'    
+        }        
+    }
+    keep if `owner_list'
```
Correct. At the moment I only left out "local government" (besides the grand total), simply because I didn't want to end up creating tooo many controls. It can be easily added though.

diegogentilepassaro replied:
I still don't understand what are the owners that we are dropping and why? Maybe a good idea is to say in a phrase what is it that we are keeping or dropping. If we are taking some particular choice, we should make a note and explain exactly what we do in the draft.

gabrieleborg replied:
we were just dropping local government, and the overall total since we are using all the industry-based breakdown. I added it now. Agree, I am going to explain that in the data section.

gabrieleborg commented on base/qcew/code/make_county_month_panel.do:
```diff
@@ -0,0 +1,100 @@
+    }
+    keep if `owner_list'
+
+    replace industry = trim(regexr(trim(industry), "^[0-9]+", ""))
+
+    local n : word count `industry'
+    local first_ind: word 1 of `industry'
+    local ind_list `"industry=="`first_ind'""'
+    
+    if `n'>1 {
+        forval x = 2/`n' {
+            local this_ind: word `x' of `industry'
+            local this_ind_list `" | industry== "`this_ind'""'
+            local ind_list `"`ind_list' `this_ind_list'"'    
+        }        
+    }
+    keep if (ownership != "Private") |  (ownership== "Private" & (`ind_list'))
+
+    replace industry = ownership if ownership!= "Private"
```
yes!

gabrieleborg commented on base/qcew/code/make_county_month_panel.do:
```diff
@@ -0,0 +1,100 @@
+
+    use `instub'/industry_county_qtr_emp_wage.dta, clear
+
+    select_sectors, ownership(`""Federal Government" "State Government" "Private""') ///
+    industry(`" "Goods-producing" "Natural resources and mining" "Construction" "Manufacturing" "Service-providing" "Trade, transportation, and utilities" "Information" "Financial Activities" "Professional and business services" "Education and health services" "Leisure and hospitality" "')
```
I spent literally 3 hours on this ahah, because ownership and industries contain multiple words names, and it was a mess to adapt locals to flexibly input them into the program. I tried by defining a local above, but it didn't work. I know it's a bit hard to read, but this was as much as I could do. Feel free to give it a try if you want, but I don't know how to do it!

santiagohermo replied:
Haha that makes sense @gabrieleborg! Probably something cleaner it's doable, but the cost is too high and we have more important thigs to do haha. If we have some time later we can try to make it prettier :)

gabrieleborg commented on derived/zipcode_month/code/create_final_panels.do:
```diff
@@ -3,19 +3,17 @@ clear all
-	} 
-
+	import delim using "`indemo'/zip_demo.csv", clear
+	save_data "`instub'/zip_ready.dta", replace key(zipcode) log(`logfile')
```
I am fine either way. Switching it to non for the moment.

gabrieleborg commented on derived/zipcode_month/code/create_final_panels.do:
```diff
@@ -57,24 +52,33 @@ program main
+	if "`bps'" == "yes" {
+		merge m:1 countyfips statefips year_month using `inbps'/bps_sf_cty_mon.dta, nogen assert(1 2 3) keep(1 3)
+	}
+end 
```
Thanks! 

I didn't use locals intentionally, because I wanted to have flexibility to decide where to add them or not. For example, we don't need QCEW and BPS data in the big general `zipcode_year_month.dta` panel, while we want them in `baseline_rent_panel.dta`. I think it's good to have such flexibility

santiagohermo replied:
That makes sense @gabrieleborg. Thanks for the clarification 👍

gabrieleborg commented on analysis/first_differences/code/first_differences.do:
```diff
@@ -11,44 +11,80 @@ program main
+	
+	esttab reg1 reg2 reg3 reg4 reg5 using "`outstub'/fd_dynamic_table_control.tex", ///
+		keep(*.ln_mw) compress se replace ///
+		stats(p_value_F employment_cov establishment_cov avg_weekly_wage_cov new_building_cov r2 N, fmt(%9.3f %s3 %s3 %s3 %s3 %9.3f %9.0gc) ///
```
good Idea! done

gabrieleborg commented on analysis/first_differences/code/first_differences.do:
```diff
@@ -11,44 +11,80 @@ program main
 	use "`instub'/fd_rent_panel.dta", clear
```
yeah that could make sense. @diegogentilepassaro what do you think?

diegogentilepassaro replied:
I think splitting it is a very good idea! Readability would be important in the near future when we create a new stable version of the repo or in the far future when we create replication files!

gabrieleborg replied:
Agree! Just to clarify: you guys want to split the main analysis (base  + controls) from the heterogeneity analysis, or you do want to split base from controls (hence having base  + heter, and controls + heter)? Because to me it seems the former, but that in the end would really shorten the code by that much since the heterogeneity part is literally 10 lines.  But it can make sense from a logical point of view.

diegogentilepassaro replied:
I agree that the code wouldn't be shortened much, but it would be something plus we will be able to play with the heterogeneity more easily ion the future! We could try to make the code for the baseline nicer and shorter in the future after we write the draft.

gabrieleborg commented on analysis/first_differences/code/first_differences.do:
```diff
@@ -11,44 +11,80 @@ program main
+	esttab * using "`outstub'/fd_table.tex", keep(D.ln_mw) compress se replace 	///
+		stats(zs_trend zs_trend_sq cty_emp_wg r2 N, fmt(%s3 %s3 %s3 %9.3f %9.0gc) ///
+		labels("Zipcode-specifc linear trend" ///
+		"Zipcode-specific linear and square trend"	///
```
done

diegogentilepassaro replied:
Sorry to bother here but I believe that "Zipcode-specific quadratic trend" is more clear.

gabrieleborg commented on analysis/first_differences/code/first_differences.do:
```diff
@@ -58,36 +94,73 @@ program run_static_model
+	syntax, depvar(str) absorb(str) cluster(str)
+
+	eststo clear
+  	local emp_cont       "ln_emp_leis ln_emp_goodpr ln_emp_const ln_emp_transp ln_emp_bizserv ln_emp_eduhe ln_emp_fedgov ln_emp_info ln_emp_manu ln_emp_natres ln_emp_servpr ln_emp_stgov"
+    local establish_cont "ln_estcount_leis ln_estcount_goodpr ln_estcount_const ln_estcount_transp ln_estcount_bizserv ln_estcount_eduhe ln_estcount_fedgov ln_estcount_info ln_estcount_manu ln_estcount_natres ln_estcount_servpr ln_estcount_stgov"
+    local wage_cont      "ln_avgwwage_leis ln_avgwwage_goodpr ln_avgwwage_const ln_avgwwage_transp ln_avgwwage_bizserv ln_avgwwage_eduhe ln_avgwwage_fedgov ln_avgwwage_info ln_avgwwage_manu ln_avgwwage_natres ln_avgwwage_servpr ln_avgwwage_stgov"
+    local housing_cont   "ln_u1rep_units ln_u1rep_value"
```
done

gabrieleborg commented on analysis/first_differences/code/first_differences.do:
```diff
@@ -179,30 +252,203 @@ program run_dynamic_model
+	test (F5D.ln_mw = 0) (F4D.ln_mw = 0) (F3D.ln_mw = 0) (F2D.ln_mw = 0) (F1D.ln_mw = 0)
+	estadd scalar p_value_F = r(p)
+
+	preserve
+		coefplot, vertical base gen
+		local winspan = 2*`w' + 1
+		keep if _n<=`winspan'
+		keep __at __b __se
+		rename (__at __b __se) (at b_emp se_emp)
+		
+		keep if !missing(at)
+
+		gen b_emp_lb = b_emp - 1.96*se_emp
+		gen b_emp_ub = b_emp + 1.96*se_emp
+		
+		save "../temp/plot_coeffs_emp.dta", replace
+	restore
```
done

gabrieleborg commented:
Thanks @santiagohermo for the review, very helpful! I updated the scripts, and I also updated the branch from master so that now the code is really complete and with no conflict.

santiagohermo commented:
The changes look good @gabrieleborg! I further removed what it seemed to be an extra tab indentation throughout the script in https://github.com/diegogentilepassaro/min_wage_rent/pull/54/commits/b8028e5bf747b4cc85b7bf1bfe6184ad056bf31b

On my side, the only thing that remains open is the splitting. I think that splitting on heterogeneity/not is intuitive, but I don't feel strongly about it. The main point for me would be that they are "balanced" in size (more than 200 lines seems too long). To make things more readable we can also send some functions to some ado file or some extra script in the same folder (e.g, the "comment_table" functions)

gabrieleborg commented:
Thanks for the feedback guys @diegogentilepassaro  @santiagohermo. 

I split the first_differences.do code into `estimate_base.do` and `estimate_heterogeneity.do`. I will go on and merge this inot master. I will then open a new quick task to "update results" in analysis by mainly running the unbalanced panel with all this econ control.
